src/net.cpp

```diff
@@ -1573,26 +1579,55 @@ void CConnman::ThreadDNSAddressSeed()
+    //   creating fewer identifying DNS requests, reduces trust by
+    //   giving seeds less influence on the network topology, and
+    //   reduces traffic to the seeds.
+    // * When querying DNS seeds query a few at once, this ensures
+    //   that we don't give DNS seeds the ability to eclipse nodes
+    //   that query them.
+    // * If we continue having problems, eventually query all the
+    //   DNS seeds, and if that fails too, also try the fixed seeds.
+    //   (done in ThreadOpenConnections)
+    const std::chrono::seconds seeds_wait_time = (addrman.size() >= DNSSEEDS_DELAY_PEER_THRESHOLD ? DNSSEEDS_DELAY_MANY_PEERS : DNSSEEDS_DELAY_FEW_PEERS);
+
     for (const std::string& seed : seeds) {
-        // goal: only query DNS seed if address need is acute
-        // Avoiding DNS seeds when we don't need them improves user privacy by
-        // creating fewer identifying DNS requests, reduces trust by giving seeds
-        // less influence on the network topology, and reduces traffic to the seeds.
         if (addrman.size() > 0 && seeds_right_now == 0) {
-            if (!interruptNet.sleep_for(std::chrono::seconds(11))) return;
-
-            LOCK(cs_vNodes);
-            int nRelevant = 0;
-            for (const CNode* pnode : vNodes) {
-                nRelevant += pnode->fSuccessfullyConnected && !pnode->fFeeler && !pnode->fOneShot && !pnode->m_manual_connection && !pnode->fInbound;
-            }
-            if (nRelevant >= 2) {
-                LogPrintf("P2P peers available. Skipped DNS seeding.\n");
-                return;
+            LogPrintf("Waiting %d seconds before querying DNS seeds.\n", seeds_wait_time.count());
+            std::chrono::seconds to_wait = seeds_wait_time;
+            while (to_wait.count() > 0) {
+                std::chrono::seconds w{25}; // check for early abort every 25 seconds
+                if (w > to_wait) w = to_wait;
+                if (!interruptNet.sleep_for(w)) return;
+                to_wait -= w;
+
+                int nRelevant = 0;
+                {
+                    LOCK(cs_vNodes);
+                    for (const CNode* pnode : vNodes) {
+                        nRelevant += pnode->fSuccessfullyConnected && !pnode->fFeeler && !pnode->fOneShot && !pnode->m_manual_connection && !pnode->fInbound;
+                    }
+                }
+                if (nRelevant >= 2) {
+                    if (found > 0) {
+                        LogPrintf("%d addresses found from DNS seeds\n", found);
+                        LogPrintf("P2P peers available. Finished DNS seeding.\n");
+                    } else {
+                        LogPrintf("P2P peers available. Skipped DNS seeding.\n");
+                    }
+                    return;
+                }
             }
             seeds_right_now += DNSSEEDS_TO_QUERY_AT_ONCE;
+        } else if (seeds_right_now == 0) {
+            seeds_right_now += DNSSEEDS_TO_QUERY_AT_ONCE;
```

Comment:
It seems this branch will never be executed since all of the dns seeds are explicitly queried if we start up with an empty addrman:
https://github.com/bitcoin/bitcoin/blob/5bf16653168cbc3c8bd0814cfa937d0dc8a879ff/src/net.cpp#L1582-L1585

Reply:
Rearranged to `if (s_r_n == 0) { s_r_n += AT_ONCE; if (addrman.size() == 0) { wait(); } }`